File: frontend/src/services/employeeService.ts
/**
 * Employee Service - API calls for employee management
 */
import apiClient from '../utils/api-client';

const BASE_URL = '/global/employees';

// ============================================
// Types
// ============================================

export type PaymentFrequency = 'weekly' | 'biweekly' | 'monthly';
export type BonusType = 'fixed' | 'variable' | 'one_time';

export interface EmployeeListItem {
  id: string;
  full_name: string;
  document_id: string;
  position: string;
  hire_date: string;
  base_salary: number;
  payment_frequency: PaymentFrequency;
  is_active: boolean;
}

export interface EmployeeCreate {
  full_name: string;
  document_type?: string;
  document_id: string;
  email?: string;
  phone?: string;
  address?: string;
  position: string;
  hire_date: string;
  base_salary: number;
  payment_frequency?: PaymentFrequency;
  payment_method?: string;
  bank_name?: string;
  bank_account?: string;
  health_deduction?: number;
  pension_deduction?: number;
  other_deductions?: number;
  user_id?: string;
}

export interface EmployeeUpdate {
  full_name?: string;
  document_type?: string;
  document_id?: string;
  email?: string;
  phone?: string;
  address?: string;
  position?: string;
  termination_date?: string;
  is_active?: boolean;
  base_salary?: number;
  payment_frequency?: PaymentFrequency;
  payment_method?: string;
  bank_name?: string;
  bank_account?: string;
  health_deduction?: number;
  pension_deduction?: number;
  other_deductions?: number;
}

export interface EmployeeResponse extends EmployeeListItem {
  document_type: string;
  email: string | null;
  phone: string | null;
  address: string | null;
  termination_date: string | null;
  payment_method: string;
  bank_name: string | null;
  bank_account: string | null;
  health_deduction: number;
  pension_deduction: number;
  other_deductions: number;
  total_deductions: number;
  user_id: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface EmployeeBonusCreate {
  name: string;
  bonus_type: BonusType;
  amount: number;
  is_recurring?: boolean;
  start_date: string;
  end_date?: string;
  notes?: string;
}

export interface EmployeeBonusUpdate {
  name?: string;
  bonus_type?: BonusType;
  amount?: number;
  is_recurring?: boolean;
  end_date?: string;
  is_active?: boolean;
  notes?: string;
}

export interface EmployeeBonusResponse {
  id: string;
  employee_id: string;
  name: string;
  bonus_type: BonusType;
  amount: number;
  is_recurring: boolean;
  start_date: string;
  end_date: string | null;
  is_active: boolean;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface EmployeeTotals {
  base_salary: number;
  total_bonuses: number;
  total_deductions: number;
  net_amount: number;
  bonus_breakdown: { name: string; amount: number }[];
  deduction_breakdown: { name: string; amount: number }[];
}

// ============================================
// API Functions
// ============================================

/**
 * List all employees
 */
export const getEmployees = async (params?: {
  skip?: number;
  limit?: number;
  is_active?: boolean;
}): Promise<EmployeeListItem[]> => {
  const response = await apiClient.get<EmployeeListItem[]>(BASE_URL, { params });
  return response.data;
};

/**
 * Get a single employee by ID
 */
export const getEmployee = async (id: string): Promise<EmployeeResponse> => {
  const response = await apiClient.get<EmployeeResponse>(`${BASE_URL}/${id}`);
  return response.data;
};

/**
 * Create a new employee
 */
export const createEmployee = async (data: EmployeeCreate): Promise<EmployeeResponse> => {
  const response = await apiClient.post<EmployeeResponse>(BASE_URL, data);
  return response.data;
};

/**
 * Update an employee
 */
export const updateEmployee = async (
  id: string,
  data: EmployeeUpdate
): Promise<EmployeeResponse> => {
  const response = await apiClient.patch<EmployeeResponse>(`${BASE_URL}/${id}`, data);
  return response.data;
};

/**
 * Delete (deactivate) an employee
 */
export const deleteEmployee = async (id: string): Promise<void> => {
  await apiClient.delete(`${BASE_URL}/${id}`);
};

/**
 * Get employee's calculated totals
 */
export const getEmployeeTotals = async (id: string): Promise<EmployeeTotals> => {
  const response = await apiClient.get<EmployeeTotals>(`${BASE_URL}/${id}/totals`);
  return response.data;
};

// ============================================
// Bonus Functions
// ============================================

/**
 * List all bonuses for an employee
 */
export const getEmployeeBonuses = async (
  employeeId: string,
  params?: { is_active?: boolean }
): Promise<EmployeeBonusResponse[]> => {
  const response = await apiClient.get<EmployeeBonusResponse[]>(
    `${BASE_URL}/${employeeId}/bonuses`,
    { params }
  );
  return response.data;
};

/**
 * Create a bonus for an employee
 */
export const createEmployeeBonus = async (
  employeeId: string,
  data: EmployeeBonusCreate
): Promise<EmployeeBonusResponse> => {
  const response = await apiClient.post<EmployeeBonusResponse>(
    `${BASE_URL}/${employeeId}/bonuses`,
    data
  );
  return response.data;
};

/**
 * Update a bonus
 */
export const updateEmployeeBonus = async (
  bonusId: string,
  data: EmployeeBonusUpdate
): Promise<EmployeeBonusResponse> => {
  const response = await apiClient.patch<EmployeeBonusResponse>(
    `${BASE_URL}/bonuses/${bonusId}`,
    data
  );
  return response.data;
};

/**
 * Delete (deactivate) a bonus
 */
export const deleteEmployeeBonus = async (bonusId: string): Promise<void> => {
  await apiClient.delete(`${BASE_URL}/bonuses/${bonusId}`);
};

// ============================================
// Helper Functions
// ============================================

/**
 * Get human-readable label for payment frequency
 */
export const getPaymentFrequencyLabel = (frequency: PaymentFrequency): string => {
  const labels: Record<PaymentFrequency, string> = {
    weekly: 'Semanal',
    biweekly: 'Quincenal',
    monthly: 'Mensual'
  };
  return labels[frequency] || frequency;
};

/**
 * Get human-readable label for bonus type
 */
export const getBonusTypeLabel = (type: BonusType): string => {
  const labels: Record<BonusType, string> = {
    fixed: 'Fijo',
    variable: 'Variable',
    one_time: 'Único'
  };
  return labels[type] || type;
};

/**
 * Format currency
 */
export const formatCurrency = (amount: number): string => {
  return `$${amount.toLocaleString('es-CO')}`;
};

export default {
  getEmployees,
  getEmployee,
  createEmployee,
  updateEmployee,
  deleteEmployee,
  getEmployeeTotals,
  getEmployeeBonuses,
  createEmployeeBonus,
  updateEmployeeBonus,
  deleteEmployeeBonus,
  getPaymentFrequencyLabel,
  getBonusTypeLabel,
  formatCurrency
};

File: backend/app/services/employee_service.py
"""
Employee Service - Business logic for employee management
"""
from uuid import UUID
from decimal import Decimal
from datetime import date
from sqlalchemy import select, and_, or_
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.payroll import Employee, EmployeeBonus, BonusType
from app.schemas.payroll import (
    EmployeeCreate,
    EmployeeUpdate,
    EmployeeBonusCreate,
    EmployeeBonusUpdate,
)


class EmployeeService:
    """Service for employee operations"""

    # ============================================
    # Employee CRUD
    # ============================================

    async def get_employees(
        self,
        db: AsyncSession,
        *,
        skip: int = 0,
        limit: int = 100,
        is_active: bool | None = None,
    ) -> list[Employee]:
        """Get all employees with optional filters"""
        stmt = select(Employee).options(selectinload(Employee.bonuses))

        if is_active is not None:
            stmt = stmt.where(Employee.is_active == is_active)

        stmt = stmt.order_by(Employee.full_name).offset(skip).limit(limit)
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def get_employee(
        self,
        db: AsyncSession,
        employee_id: UUID,
    ) -> Employee | None:
        """Get a single employee by ID"""
        stmt = (
            select(Employee)
            .options(selectinload(Employee.bonuses))
            .where(Employee.id == employee_id)
        )
        result = await db.execute(stmt)
        return result.scalar_one_or_none()

    async def get_employee_by_document(
        self,
        db: AsyncSession,
        document_id: str,
    ) -> Employee | None:
        """Get employee by document ID"""
        stmt = select(Employee).where(Employee.document_id == document_id)
        result = await db.execute(stmt)
        return result.scalar_one_or_none()

    async def create_employee(
        self,
        db: AsyncSession,
        data: EmployeeCreate,
        *,
        created_by: UUID | None = None,
    ) -> Employee:
        """Create a new employee"""
        # Check for duplicate document_id
        existing = await self.get_employee_by_document(db, data.document_id)
        if existing:
            raise ValueError(f"Ya existe un empleado con el documento {data.document_id}")

        employee = Employee(
            **data.model_dump(),
            created_by=created_by,
        )
        db.add(employee)
        await db.commit()
        await db.refresh(employee)
        return employee

    async def update_employee(
        self,
        db: AsyncSession,
        employee_id: UUID,
        data: EmployeeUpdate,
    ) -> Employee:
        """Update an employee"""
        employee = await self.get_employee(db, employee_id)
        if not employee:
            raise ValueError("Empleado no encontrado")

        update_data = data.model_dump(exclude_unset=True)

        # Check for document_id uniqueness if being updated
        if "document_id" in update_data and update_data["document_id"] != employee.document_id:
            existing = await self.get_employee_by_document(db, update_data["document_id"])
            if existing:
                raise ValueError(f"Ya existe un empleado con el documento {update_data['document_id']}")

        for field, value in update_data.items():
            setattr(employee, field, value)

        await db.commit()
        await db.refresh(employee)
        return employee

    async def delete_employee(
        self,
        db: AsyncSession,
        employee_id: UUID,
    ) -> bool:
        """Soft delete an employee (set is_active to False)"""
        employee = await self.get_employee(db, employee_id)
        if not employee:
            return False

        employee.is_active = False
        employee.termination_date = date.today()
        await db.commit()
        return True

    # ============================================
    # Employee Bonus CRUD
    # ============================================

    async def get_employee_bonuses(
        self,
        db: AsyncSession,
        employee_id: UUID,
        *,
        is_active: bool | None = None,
    ) -> list[EmployeeBonus]:
        """Get all bonuses for an employee"""
        stmt = select(EmployeeBonus).where(EmployeeBonus.employee_id == employee_id)

        if is_active is not None:
            stmt = stmt.where(EmployeeBonus.is_active == is_active)

        stmt = stmt.order_by(EmployeeBonus.name)
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def get_bonus(
        self,
        db: AsyncSession,
        bonus_id: UUID,
    ) -> EmployeeBonus | None:
        """Get a single bonus by ID"""
        stmt = select(EmployeeBonus).where(EmployeeBonus.id == bonus_id)
        result = await db.execute(stmt)
        return result.scalar_one_or_none()

    async def create_bonus(
        self,
        db: AsyncSession,
        employee_id: UUID,
        data: EmployeeBonusCreate,
    ) -> EmployeeBonus:
        """Create a bonus for an employee"""
        # Verify employee exists
        employee = await self.get_employee(db, employee_id)
        if not employee:
            raise ValueError("Empleado no encontrado")

        bonus = EmployeeBonus(
            employee_id=employee_id,
            **data.model_dump(),
        )
        db.add(bonus)
        await db.commit()
        await db.refresh(bonus)
        return bonus

    async def update_bonus(
        self,
        db: AsyncSession,
        bonus_id: UUID,
        data: EmployeeBonusUpdate,
    ) -> EmployeeBonus:
        """Update a bonus"""
        bonus = await self.get_bonus(db, bonus_id)
        if not bonus:
            raise ValueError("Bono no encontrado")

        update_data = data.model_dump(exclude_unset=True)
        for field, value in update_data.items():
            setattr(bonus, field, value)

        await db.commit()
        await db.refresh(bonus)
        return bonus

    async def delete_bonus(
        self,
        db: AsyncSession,
        bonus_id: UUID,
    ) -> bool:
        """Soft delete a bonus"""
        bonus = await self.get_bonus(db, bonus_id)
        if not bonus:
            return False

        bonus.is_active = False
        await db.commit()
        return True

    # ============================================
    # Helper Methods
    # ============================================

    async def get_active_bonuses_for_date(
        self,
        db: AsyncSession,
        employee_id: UUID,
        target_date: date,
    ) -> list[EmployeeBonus]:
        """Get active bonuses for an employee on a specific date"""
        stmt = select(EmployeeBonus).where(
            and_(
                EmployeeBonus.employee_id == employee_id,
                EmployeeBonus.is_active == True,
                EmployeeBonus.start_date <= target_date,
                or_(
                    EmployeeBonus.end_date == None,
                    EmployeeBonus.end_date >= target_date,
                ),
            )
        )
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def calculate_employee_totals(
        self,
        db: AsyncSession,
        employee_id: UUID,
        target_date: date | None = None,
    ) -> dict:
        """Calculate total bonuses and deductions for an employee"""
        if target_date is None:
            target_date = date.today()

        employee = await self.get_employee(db, employee_id)
        if not employee:
            raise ValueError("Empleado no encontrado")

        # Get active bonuses
        bonuses = await self.get_active_bonuses_for_date(db, employee_id, target_date)

        total_bonuses = sum(b.amount for b in bonuses if b.is_recurring or b.bonus_type == BonusType.ONE_TIME)
        total_deductions = (
            employee.health_deduction +
            employee.pension_deduction +
            employee.other_deductions
        )

        bonus_breakdown = [
            {"name": b.name, "amount": float(b.amount)}
            for b in bonuses
            if b.is_recurring or b.bonus_type == BonusType.ONE_TIME
        ]

        deduction_breakdown = []
        if employee.health_deduction > 0:
            deduction_breakdown.append({"name": "Salud", "amount": float(employee.health_deduction)})
        if employee.pension_deduction > 0:
            deduction_breakdown.append({"name": "Pensión", "amount": float(employee.pension_deduction)})
        if employee.other_deductions > 0:
            deduction_breakdown.append({"name": "Otras deducciones", "amount": float(employee.other_deductions)})

        net_amount = employee.base_salary + total_bonuses - total_deductions

        return {
            "base_salary": employee.base_salary,
            "total_bonuses": total_bonuses,
            "total_deductions": total_deductions,
            "net_amount": net_amount,
            "bonus_breakdown": bonus_breakdown,
            "deduction_breakdown": deduction_breakdown,
        }


# Create singleton instance
employee_service = EmployeeService()
